packages/pipeline/src/data_sources/ohlcv_external/crypto_compare.ts

```diff
@@ -0,0 +1,88 @@
+// tslint:disable:no-duplicate-imports
+import { AxiosResponse } from 'axios';
+import axios from 'axios';
+import promiseLimit = require('p-limit');
+
+import { merge, unfold } from 'ramda';
+
+import { TradingPair } from '../../utils/get_ohlcv_trading_pairs';
+
+export interface CryptoCompareOHLCVResponse {
+    Data: CryptoCompareOHLCVRecord[];
+}
+
+export interface CryptoCompareOHLCVRecord {
+    time: number; // in seconds, not milliseconds
+    close: number;
+    high: number;
+    low: number;
+    open: number;
+    volumefrom: number;
+    volumeto: number;
+}
+
+export interface CryptoCompareOHLCVParams {
+    fsym: string;
+    tsym: string;
+    e?: string;
+    aggregate?: string;
+    aggregatePredictableTimePeriods?: boolean;
+    limit?: number;
+    toTs?: number;
+}
+
+// tslint:disable:custom-no-magic-numbers
+const ONE_WEEK = 7 * 24 * 60 * 60 * 1000;
+const ONE_SECOND = 1000;
+
+const HISTORICAL_HOURLY_OHLCV_URI = 'https://min-api.cryptocompare.com/data/histohour';
+
+export class CryptoCompareOHLCVSource {
+    public readonly interval = ONE_WEEK; // the hourly API returns data for one week at a time
+    public readonly default_exchange = 'CCCAGG';
+    private readonly _url: string = HISTORICAL_HOURLY_OHLCV_URI;
+    private readonly _plimit: (
+        fetchFn: () => Promise<AxiosResponse<CryptoCompareOHLCVResponse>>,
+    ) => Promise<AxiosResponse<CryptoCompareOHLCVResponse>>;
+
+    constructor(maxConcurrentRequests: number = 50) {
+        this._plimit = promiseLimit(maxConcurrentRequests);
+    }
+
+    // gets OHLCV records starting from pair.latest
+    public async getAsync(pair: TradingPair): Promise<CryptoCompareOHLCVRecord[]> {
+        const params = {
+            e: this.default_exchange,
+            fsym: pair.fromSymbol,
+            tsym: pair.toSymbol,
+            toTs: Math.floor((pair.latest + this.interval) / ONE_SECOND), // CryptoCompare uses timestamp in seconds. not ms
+        };
+
+        const fetchPromise = this._plimit(() => {
+            // tslint:disable:no-console
+            console.log(`Scraping Crypto Compare with ${JSON.stringify(params)}`);
+            return axios.get<CryptoCompareOHLCVResponse>(this._url, { params });
+        });
+
+        const resp = await Promise.resolve(fetchPromise);
```

Comment:
I'm referencing a single instance of the plimit function as a private member scoped to the whole class, so that when we instantiate this class it enforces rate limiting for all reqs to the API.

I think your understanding is correct. `this._plimit(() => asyncFunction)` returns a promise wrapping the result of async function. The plimit example code is structured as:
```
Promise.all([
    plimit(fn1),
    plimit(fn2),
    plimit(fn3),
...])
```

But since I'm only calling one function here, I'm using `Promise.resolve` instead of `Promise.all`. 

Does that make sense? I'm open to renaming stuff to make the code more understandable too

Reply:
@xianny I think renaming some things and possibly adding some comments would greatly help comprehensibility. Most people on the team will have never encountered `p-limit` before.

Reply:
I added some type annotations and comments :)